File: sem-2/Assignment/List_Python_in4L_assignment_1/1.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class List {
private:
    vector<int> array;

public:
    void append(int x) {
        array.push_back(x);
    }

    void extend(vector<int>& iterable) {
        array.insert(array.end(), iterable.begin(), iterable.end());
    }

    void insert(int index, int x) {
        if (index < 0 || index > array.size())
            return;  
        array.insert(array.begin() + index, x);
    }

    void remove(int x) {
        array.erase(remove(array.begin(), array.end(), x), array.end());
    }

    int pop(int index) {
        if (index < 0 || index >= array.size())
            return -1;  
        int popped = array[index];
        array.erase(array.begin() + index);
        return popped;
    }

    void clear() {
        array.clear();
    }

    int index(int x, int start = 0, int end = -1) {
        if (end == -1)
            end = array.size();
        for (int i = start; i < end; i++) {
            if (array[i] == x)
                return i;
        }
        return -1;  
    }

    int count(int x) {
        return std::count(array.begin(), array.end(), x);
    }

    void sort(bool reverse = false) {
        if (reverse)
            std::sort(array.rbegin(), array.rend());
        else
            std::sort(array.begin(), array.end());
    }

    void reverse() {
        std::reverse(array.begin(), array.end());
    }

    vector<int> copy() {
        return array;
    }

    void print() {
        cout << "[";
        for (size_t i = 0; i < array.size(); i++) {
            cout << array[i];
            if (i != array.size() - 1)
                cout << ", ";
        }
        cout << "]" << endl;
    }
};

int main() {
    List myList;
    myList.append(1);
    myList.append(2.2);
    myList.append(3);
    myList.append(4);
    cout << "Original List: ";
    myList.print();
    myList.insert(2, 10);
    cout << "After Insertion: ";
    myList.print();
    myList.remove(3);
    cout << "After Removal: ";
    myList.print();
    cout << "Popped: " << myList.pop(1) << endl;
    cout << "After Pop: ";
    myList.print();
    myList.sort();
    cout << "After Sorting: ";
    myList.print();
    myList.reverse();
    cout << "After Reversing: ";
    myList.print();
    vector<int> copiedList = myList.copy();
    cout << "Copied List: ";
    for (int num : copiedList)
        cout << num << " ";
    cout << endl;
    myList.clear();
    cout << "After Clearing: ";
    myList.print();
    return 0;
}

File: sem-2/Assignment/List_Python_in4L_assignment_1/1.js


function List() {
    this.array = [];
}

List.prototype.append = function(x) {
    this.array.push(x);
};

List.prototype.extend = function(iterable) {
    this.array.push(...iterable);
};

List.prototype.insert = function(index, x) {
    if (index < 0 || index > this.array.length)
        return;  
    this.array.splice(index, 0, x);
};

List.prototype.remove = function(x) {
    const index = this.array.indexOf(x);
    if (index !== -1)
        this.array.splice(index, 1);
};

List.prototype.pop = function(index) {
    if (index < 0 || index >= this.array.length)
        return -1;  
    return this.array.splice(index, 1)[0];
};

List.prototype.clear = function() {
    this.array = [];
};

List.prototype.index = function(x, start = 0, end = this.array.length) {
    if (start < 0)
        start = 0;
    if (end > this.array.length)
        end = this.array.length;
    return this.array.slice(start, end).indexOf(x);
};

List.prototype.count = function(x) {
    return this.array.filter(num => num === x).length;
};

List.prototype.sort = function(reverse = false) {
    this.array.sort((a, b) => reverse ? b - a : a - b);
};

List.prototype.reverse = function() {
    this.array.reverse();
};

List.prototype.copy = function() {
    return [...this.array];
};

List.prototype.print = function() {
    console.log(`[${this.array.join(', ')}]`);
};

// Test
const myList = new List();
myList.append(1);
myList.append(2);
myList.append(3);
myList.append(4);
console.log("Original List: ");
myList.print();
myList.insert(2, 10);
console.log("After Insertion: ");
myList.print();
myList.remove(3);
console.log("After Removal: ");
myList.print();
console.log("Popped: " + myList.pop(1));
console.log("After Pop: ");
myList.print();
myList.sort();
console.log("After Sorting: ");
myList.print();
myList.reverse();
console.log("After Reversing: ");
myList.print();
const copiedList = myList.copy();
console.log("Copied List: ", copiedList);
myList.clear();
console.log("After Clearing: ");
myList.print();
